blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+1013  /buildworker/worker/package_linux64/build/src/jltypes.c: No such file or directory.
+
+typekey_eq (key=key@entry=0x435f2dc8, n=n@entry=46200, tt=<optimized out>, tt=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/jltypes.c:560
+560 in /buildworker/worker/package_linux64/build/src/jltypes.c
+(rr) 
+Continuing.
+[FATAL /home/keno/rr/src/ReplaySession.cc:636:check_pending_sig()] 
+ (task 19490 (rec:7760) at time 51911)
+ -> Assertion `false' failed to hold. Replaying `SIGNAL: SIGSEGV(det)': expecting tracee signal or trap, but instead at `write' (ticks: 144040993972)
+```
+
+We see some accesses from the type cache, and then the divergence. What about in the other direction? (Note that the divergence crashed the rr process, so we have to first get back to the same execution point - luckily since this is a recording, everything is stable, so we can just `seek-ticks` right back to where we were).
+
+```
+(rr) awatch *(void**)0x436487c0
+Hardware access (read/write) watchpoint 1: *(void**)0x436487c0
+(rr) rc
+Continuing.
+
+Thread 1 hit Hardware access (read/write) watchpoint 1: *(void**)0x436487c0
+
+Old value = (void *) 0x7f0490b784d0 <jl_system_image_data+16912>
+New value = (void *) 0x0
+jl_svecset (x=0x7f0490b784d0 <jl_system_image_data+16912>, i=<optimized out>, t=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/julia.h:840
+840 /buildworker/worker/package_linux64/build/src/julia.h: No such file or directory.
+(rr) bt
+#0  jl_svecset (x=0x7f0490b784d0 <jl_system_image_data+16912>, i=<optimized out>, t=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/julia.h:840
+#1  jl_f_tuple (F=<optimized out>, args=0x42f54590, nargs=46200)
+    at /buildworker/worker/package_linux64/build/src/builtins.c:728
+```
+
+For reference (from the 1.5 release branch):
+
+```
+JL_CALLABLE(jl_f_tuple)
+{
+    size_t i;
+    if (nargs == 0)
+        return (jl_value_t*)jl_emptytuple;
+    jl_datatype_t *tt;
+    if (nargs < jl_page_size / sizeof(jl_value_t*)) {
+        jl_value_t **types = (jl_value_t**)alloca(nargs * sizeof(jl_value_t*));
+        for (i = 0; i < nargs; i++)
+            types[i] = jl_typeof(args[i]);
+        tt = jl_inst_concrete_tupletype_v(types, nargs);
+    }
+    else {
+        jl_svec_t *types = jl_alloc_svec_uninit(nargs);
+        JL_GC_PUSH1(&types);
+        for (i = 0; i < nargs; i++)
+            jl_svecset(types, i, jl_typeof(args[i]));
+        tt = jl_inst_concrete_tupletype(types);
+        JL_GC_POP();
+    }
+    if (tt->instance != NULL)
+        return tt->instance;
+    jl_ptls_t ptls = jl_get_ptls_states();
+    jl_value_t *jv = jl_gc_alloc(ptls, jl_datatype_size(tt), tt);
+    for (i = 0; i < nargs; i++)
+        set_nth_field(tt, (void*)jv, i, args[i]);
+    return jv;
+}
+```
+
+So we're just getting the type pointer from some array. We can find out which one:
```

Comment:
Instantly lost me.  😂 

I'm not sure what the C code is supposed to illustrate; if it's supposed to communicate to us that the code is straightforward, it may not be doing its job.  Perhaps you can shrink it down to 2-3 lines?  I can read a backtrace, so I know that it's focused on `jl_svecset`, of which there is only one in the above function, but there's a lot going on in that function and none of it is commented, so the casual reader will most likely either completely skip this or be confused by it.

It's also not clear what "getting the type pointer from some array" means.  If you could unpack that a bit, (e.g. "getting a pointer to the type information associated with this array, a common operation done by the Julia runtime anytime it needs to access the type information of an object") it would also be helpful.